Note removed Util modules.

gguillotte opened the pull request:
Supports DOC-2401.

Looks like BBEdit also stripped some trailing spaces (3.0 L79, 93; 3.1 L34).

CC @jtappa @kylog @MikaelSmith

kylog commented on source/facter/3.0/release_notes.md:
```diff
@@ -235,6 +235,18 @@ We changed the value of the hardware fact (`os.hardware`) from "x64" to "x86_64"
 
+### BREAK: Removed `file_read.rb` and `registry.rb`
+
+Facter 3 removed the [`file_read.rb`](https://github.com/puppetlabs/facter/blob/2.x/lib/facter/util/file_read.rb) and [`registry.rb`](https://github.com/puppetlabs/facter/blob/2.x/lib/facter/util/registry.rb), which provide the `Facter::Util::FileRead` and `Facter::Util::Registry` modules to Facter. This can break custom facts that rely on these modules, and produce error messages like this:
```
IIUC the issue isn't specific to these two files. E.g. look at https://github.com/puppetlabs/facter/tree/2.x/lib/facter/util, none of which exist in facter 3. Thus, the reimplementation of facter from ruby to C++ means custom facts require'ing any individual file in the previous implementation will fail. 

In addition, whereas facter 2 effectively exposed almost its entire implementation as if it were public API for authors of custom facts, facter 3 supports a well-considered API: https://github.com/puppetlabs/facter/blob/master/Extensibility.md#custom-facts-compatibility. Thus, custom facts using facter 2 "API"s outside the facter 3 supported set will fail.

With that as background, I'd suggest that:
* we document the APIs in that Extensibility.md link
* this section be reworked to instruct custom facts authors not to require individual implementation files and to prefer only the documented APIs

Since that first point may take a bit of time we could just link to the Extensibility.md for the time being.

And then that second point could then be enhanced to say something like: if you are porting custom facts from facter 2 that rely on old APIs, a workaround would be to copy the facter 2 implementation into the module's `lib/facter` directory as  described below.

Hope that makes sense. Happy to chat more about this in meatspace.

Also /cc @peterhuene for his thoughts.

gguillotte commented:
Thanks, @kylog! I've revised the notes. CC @jtappa @peterhuene

kylog commented on source/facter/3.0/release_notes.md:
```diff
@@ -235,6 +235,18 @@ We changed the value of the hardware fact (`os.hardware`) from "x64" to "x86_64"
+
+As a workaround, you can rewrite your custom facts to avoid requiring these Ruby implementations, and instead use the [new API](https://github.com/puppetlabs/facter/blob/master/Extensibility.md#custom-facts-compatibility). You can also [copy the necessary Facter 2 Ruby libraries](https://github.com/puppetlabs/facter/tree/2.x/lib/facter/util) to the `lib/facter/util/` subdirectory of the affected  ...
+
+* [PE-13225](https://tickets.puppetlabs.com/browse/PE-13225)
```
This was an internal project ticket, so I just moved it to https://tickets.puppetlabs.com/browse/FACT-1400.

petems commented:
@gguillotte Awesome, I've had this bite a few times with customers using `Facter::Util::IP` in custom facts during an upgrade 👍

gguillotte commented:
Updated the ticket numbers per @kylog.

CC @jtappa @nfagerlund for a docs review.

nfagerlund commented on source/facter/3.1/release_notes.md:
```diff
@@ -282,3 +282,14 @@ The `puppet-agent` package that installs Facter 3.0.0 did not create the `rubysi
 
 * [FACT-1154](https://tickets.puppetlabs.com/browse/FACT-1154)
 
+### BREAK: Removed Ruby Facter implementations
```
I think this shouldn't be in the Facter 3.1 release notes, since the behavior was changed in 3.0. I'd be :+1: on putting a note in the Facter 3.1 custom facts guides, but :-1: to putting it in the release notes.

nfagerlund commented:
Other than where to say it, this lgtm.

gguillotte commented:
- Removed the note from the 3.1 release notes
- At the top of the 3.1 notes, emphasized checking the 3.0 notes for breaks
- Added short versions of the RN to the 3.0 and 3.1 `custom_facts` doc

nfagerlund commented:
:shipit:?

gguillotte commented:
@nfagerlund Fine by me
